Battle Engine Upstream Sync

ultima-soul opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
First merge details:
Conflicts in:
```
asm/macros/battle_script.inc: jumpifability, various7, various14, various15, various24, setstatchanger, setmoveeffect, jumpifstatus3
asm/macros/event.inc: braillemsgbox was added after dofieldeffectsparkle
data/battle_anim_scripts.s: Move_MUDDY_WATER, SetSolarbeamBg -> SetSolarBeamBg, AnimTask_CreateSmallSolarbeamOrbs -> AnimTask_CreateSmallSolarBeamOrbs
data/battle_scripts_1.s: gBattleScriptsForMoveEffects, BattleScript_ExplosionLoop, BattleScript_MoveMissedDoDamage, BattleScript_EffectSolarBeam, BattleScript_HandleFaintedMon, BattleScript_PursuitDmgOnSwitchOut, BattleScript_DamagingWeatherContinues, BattleScript_DmgHazardsOnAttacker, BattleScript_DmgHazardsOnTarget, BattleScript_DmgHazardsOnFaintedBattler, BattleScript_GulpMissileGorgingTargetDefenseCantGoLower, BattleScript_DoTurnDmg, BattleScript_DoRecoil, BattleScript_DrizzleActivates, BattleScript_RoughSkinActivates, BattleScript_SpikyShieldEffect, BattleScript_MoveUsedLoafingAround, BattleScript_ItemHealHP_RemoveItemEnd2, BattleScript_BerryPPHealEnd2, BattleScript_SelectingNotAllowedMoveGorillaTactics, BattleScript_BerryConfuseHealEnd2, BattleScript_BerryStatRaiseEnd2, Various scripts `tryfaintmon BS_*, FALSE, NULL` -> `tryfaintmon BS_*`, BattleScript_BadDreamsLoop `atk24` -> `checkteamslost`, BattleScript_DancerActivates `HITMARKER_x800000` -> `HITMARKER_ALLOW_NO_PP`, Various scripts `playanimation BS_*, B_ANIM_*, NULL` -> `playanimation BS_*, B_ANIM_*`, BattleScript_EffectGrowth `WEATHER_SUN_ANY` -> `B_WEATHER_SUN`, Various scripts for weather constants
include/battle.h: struct SpecialStatus, struct BattleStruct, struct BattleScripting -> field_20 was renamed in Dancer PR to savedBattler while pret renamed to pursuitDoublesAttacker
include/battle_controllers.h
include/battle_util.h
include/constants/battle.h
include/constants/battle_anim.h
include/constants/battle_move_effects.h
include/constants/battle_script_commands.h
include/constants/pokemon.h
include/graphics.h
include/pokemon.h
deleted by us:   src/battle_ai_script_commands.c
src/battle_ai_switch_items.c: ShouldSwitchIfWonderGuard
src/battle_anim.c: LaunchBattleAnimation, Cmd_end
src/battle_anim_effects_1.c: AnimSolarBeamBigOrb, AnimSolarBeamSmallOrb
src/battle_anim_effects_3.c: AnimTask_GetWeather
src/battle_anim_electric.c: CreateShockWaveLightningSprite
src/battle_anim_normal.c: AnimTask_TintPalettes
src/battle_controller_opponent.c: Intro_TryShinyAnimShowHealthbox, OpponentHandleGetRawMonData, OpponentHandleChooseMove
src/battle_controller_player.c: HandleInputChooseTarget, HandleInputChooseMove, PlayerHandleGetRawMonData, PlayerHandleReturnMonToBall
src/battle_controller_player_partner.c: PlayerPartnerHandleChooseMove
src/battle_controllers.c: BtlController_EmitChoosePokemon
src/battle_gfx_sfx_util.c: ChooseMoveAndTargetInBattlePalace, BattleLoadMonSpriteGfx, HandleSpeciesGfxDataChange
src/battle_interface.c: Deleted data returned due to modification in pret, UpdateLvlInHealthbox, UpdateHpTextInHealthbox
src/battle_main.c: sTurnActionsFuncsTable, CB2_InitBattleInternal, SpriteCB_Flicker, FaintClearSetData, DoBattleIntro, TryDoEventsBeforeFirstTurn, BattleTurnPassed, HandleTurnActionSelectionState, UpdateBattlerPartyOrdersOnSwitch, GetBattlerTotalSpeedStat, RunTurnActionsFunctions, Documentation for Cmd_hiddenpowercalc ported to SetTypeBeforeUsingMove, Documentation for Cmd_setweatherballtype ported to SetTypeBeforeUsingMove
src/battle_message.c: SetPpNumbersPaletteInMoveSelection
src/battle_script_commands.c: sOamData_MonIconOnLvlUpBox, Cmd_attackcanceler, AccuracyCalcHelper, Cmd_accuracycheck, Cmd_datahpupdate, Cmd_effectivenesssound, Cmd_resultmessage, SetMoveEffect, Cmd_seteffectwithchance, Cmd_tryfaintmon, NoAliveMonsForPlayer, NoAliveMonsForOpponent, Cmd_checkteamslost, Cmd_playanimation, Cmd_playanimation_var, Cmd_moveend, Cmd_switchindataupdate, CanBattlerSwitch, Cmd_openpartyscreen, Cmd_switchhandleorder, Cmd_switchineffects, Cmd_faintifabilitynotdamp, Cmd_setrain, ChangeStatBuffs, Cmd_setsandstorm, Cmd_psywavedamageeffect, Cmd_settypetorandomresistance, IsTwoTurnsMove, AttacksThisTurn, Cmd_setsunny, Cmd_rapidspinfree, Cmd_recoverbasedonsunlight, Cmd_setstickyweb, Cmd_sethail, Cmd_callterrainattack, Cmd_tryswapitems, DoesSubstituteBlockMove, Cmd_tryrecycleitem, Cmd_pursuitdoubles -> field_20 was renamed in Dancer PR to savedBattler while pret renamed to pursuitDoublesAttacker, Cmd_removelightscreenreflect, HandleBattleWindow, WEATHER constants adjusted
src/battle_tv.c: sPoints_MoveEffect
src/battle_util.c: HandleAction_UseMove, HandleAction_NothingIsFainted, HandleAction_ActionFinished, // rom const data being deleted in pret, TrySetCantSelectMoveBattleScript, DoFieldEndTurnEffects, DoBattlerEndTurnEffects (cases ENDTURN_ITEMS1, ENDTURN_ITEMS2, ENDTURN_UPROAR, ENDTURN_THRASH, ENDTURN_YAWN), HandleFaintedMonActions, TryClearRageAndFuryCutter, AtkCanceller_UnableToUseMove (cases CANCELLER_FLINCH), TryWeatherFormChange, AbilityBattleEffects (cases ABILITYEFFECT_ON_SWITCHIN: ABILITYEFFECT_SWITCH_IN_WEATHER, ABILITY_DRIZZLE, ABILITY_SAND_STREAM, ABILITY_DROUGHT, ABILITY_FORECAST; ABILITYEFFECT_ENDTURN: ABILITY_RAIN_DISH, ABILITY_SHED_SKIN; ABILITYEFFECT_IMMUNITY: ABILITY_INSOMNIA; ABILITYEFFECT_SYNCHRONIZE; ABILITYEFFECT_ATK_SYNCHRONIZE; ABILITYEFFECT_INTIMIDATE1; ABILITYEFFECT_TRACE), ItemBattleEffects, ClearFuryCutterDestinyBondGrudge, GetMoveTarget, IsMonDisobedient
src/contest_effect.c: AreMovesContestCombo
src/graphics.c: gUnknownGfx_C06D98 -> gBattleAnimSpriteGfx_SpinningBall area, gBattleAnimBgPalette_Solarbeam -> gBattleAnimBgPalette_SolarBeam area + gUnknown_E6BC04 -> gBattleAnimBgTilemap_SolarBeam
src/item_use.c: ItemUseInBattle_PokeBall
src/pokemon.c: SetWildMonHeldItem
```
Also edited:
```
src/battle_ai_main.c: EFFECT_SOLARBEAM -> EFFECT_SOLAR_BEAM, 0xFF -> MOVE_LIMITATIONS_ALL for CheckMoveLimitations, WEATHER constants adjusted
src/data/battle_anim.h: Ported BE changes and adjusted formatting (this was a new file from pret)
include/battle_anim.h: AnimSolarbeamBigOrb -> AnimSolarBeamBigOrb, gSolarbeamBigOrbAnimTable -> gSolarBeamBigOrbAnimTable
src/battle_anim_new.c: AnimSolarbeamBigOrb -> AnimSolarBeamBigOrb, gSolarbeamBigOrbAnimTable -> gSolarBeamBigOrbAnimTable
src/battle_ai_util.c: EFFECT_SOLARBEAM -> EFFECT_SOLAR_BEAM, WEATHER constants adjusted
src/data/battle_moves.h: EFFECT_SOLARBEAM -> EFFECT_SOLAR_BEAM
src/battle_debug.c: Removed parentheses where appropriate for ~ usage
```
Second merge details:
Conflicts in:
```
include/constants/battle.h
src/battle_anim_effects_2.c
src/battle_anim_effects_3.c: AnimTask_TransformMon
src/battle_interface.c
src/battle_main.c: DoBattleIntro
```
Also edited:
```
src/battle_anim_new.c: SpriteCB_GrowingSuperpower, AnimHappyHourCoinShower
src/battle_anim_ice.c: AvalancheAnim_Step, AvalancheAnim_Step2
```
Third merge details:
Conflicts in:
```
.github/workflows/build.yml
asm/macros/battle_ai_script.inc: if_move_flag, if_field_status, get_move_accuracy, call_if_eq, call_if_move_flag
asm/macros/event.inc: setwildbattle
include/global.h: includes at top
src/battle_message.c: HANDLE_NICKNAME_STRING_CASE, BattleStringExpandPlaceholders, ExpandBattleTextBuffPlaceholders
src/pokemon.c: PokemonToBattleMon
src/scrcmd.c: global variable declarations at top
```
Latest BE commit merge details:
Conflicts in:
```
include/battle_util.h: ABILITYEFFECT_* constants
src/battle_ai_main.c: AI_CheckBadMove
src/battle_ai_switch_items.c: FindMonWithFlagsAndSuperEffective, AI_TrySwitchOrUseItem
src/battle_controller_opponent.c: OpponentHandleChooseMove
src/battle_controller_player_partner.c: PlayerPartnerHandleChooseMove
src/battle_main.c: HandleTurnActionSelectionState
src/battle_script_commands.c: Cmd_switchineffects, Cmd_various, Cmd_switchoutabilities
src/battle_util.c: TryWeatherFormChange
```
## **Discord contact info**
<!--- formatted as name#numbers, e.g. PikalaxALT#5823 -->
<!--- Contributors must join https://discord.gg/d5dubZ3 -->
UltimaSoul#4017

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -1776,14 +1768,14 @@ u8 TrySetCantSelectMoveBattleScript(void)
         }
     }
-    if ((GetBattlerAbility(gActiveBattler) == ABILITY_GORILLA_TACTICS) && *choicedMove != 0 
+    if ((GetBattlerAbility(gActiveBattler) == ABILITY_GORILLA_TACTICS) && *choicedMove != MOVE_NONE 
```
```suggestion
    if ((GetBattlerAbility(gActiveBattler) == ABILITY_GORILLA_TACTICS) && *choicedMove != MOVE_NONE
```

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -3765,22 +3774,22 @@ u8 TryWeatherFormChange(u8 battler)
+        else if (gBattleWeather & B_WEATHER_HAIL && !IS_BATTLER_OF_TYPE(battler, TYPE_ICE))
         {
             SET_BATTLER_TYPE(battler, TYPE_ICE);
             ret = 4;
```
`ret` assignment should use the castform constants introduced in `inclunde/constants/battle.h`
(`CASTFORM_NORMAL`, `CASTFORM_FIRE`, etc.).
Should also use the new Cherrim constants in its respective section.

AsparagusEduardo reviewed (changes requested):
Other comments:
- Calls to `CheckMoveLimitations` have been updated so when they use the last parameter as `MOVE_LIMITATIONS_ALL` instead of `0xFF`. `src/battle_ai_util.c` has 6 instances where this needs to be updated.
- In `NoTargetPresent` in `src/battle_script_commands.c`
  - `gBattlerTarget = GetMoveTarget(move, 0);` should be changed to `gBattlerTarget = GetMoveTarget(move, NO_TARGET_OVERRIDE);`.
- In `Cmd_various` in `case VARIOUS_MOVEEND_ITEM_EFFECTS:`
  - `if (ItemBattleEffects(1, gActiveBattler, FALSE))` should be changed to `if (ItemBattleEffects(ITEMEFFECT_NORMAL, gActiveBattler, FALSE))`
- In `AbilityBattleEffects`:
  - When checking for `ABILITY_PASTEL_VEIL`, `gSpecialStatuses[battler].switchInAbilityDone = 1;` should be changed to `gSpecialStatuses[battler].switchInAbilityDone = TRUE;`

ultima-soul commented:
Updated with details of second and third pret merge plus merge with latest BE commits from yesterday. Review suggestions will be addressed after the review of these additions.

AsparagusEduardo reviewed (changes requested):
Here's the comparison between our branches with the new changes from pret + changes from BE:
https://github.com/rh-hideout/pokeemerald-expansion/compare/7369f60..eea3cf6

Excluding the changes suggested in the first review, here are my comments:
- In `include/battle_gfx_sfx_util.h`, `HandleSpeciesGfxDataChange` has the argument `notTransform`, but in `src/battle_gfx_sfx_util.c` it's called `castform`.
- In `FindMonWithFlagsAndSuperEffective`, the line `BtlController_EmitTwoReturnValues(BUFFER_B, B_ACTION_SWITCH, 0);` was deleted in #1954.
- I have modified `SetWildMonHeldItem` to pret's standard while also applying the double battle changes needed by it, instead of keeping it how it was before. This way, we're on top of the documentation.

ultima-soul commented:
> Here's the comparison between our branches with the new changes from pret + changes from BE: https://github.com/rh-hideout/pokeemerald-expansion/compare/7369f60..eea3cf6
> 
> Excluding the changes suggested in the first review, here are my comments:
> 
> * In `include/battle_gfx_sfx_util.h`, `HandleSpeciesGfxDataChange` has the argument `notTransform`, but in `src/battle_gfx_sfx_util.c` it's called `castform`.
> * In `FindMonWithFlagsAndSuperEffective`, the line `BtlController_EmitTwoReturnValues(BUFFER_B, B_ACTION_SWITCH, 0);` was deleted in #1954.
> * I have modified `SetWildMonHeldItem` to pret's standard while also applying the double battle changes needed by it, instead of keeping it how it was before. This way, we're on top of the documentation.

Point 1: This should be fixed in pret along with the other two argument names.
Point 2: It was deleted in the revamp, but then in the reversion, it was forgotten to be added back so I added it back.
